File: scripts/test_solutions/207_course_schedule.js
var canFinish = function(numCourses, prerequisites) {
    const graph = Array(numCourses).fill(null).map(() => []);
    const indegree = Array(numCourses).fill(0);
    for (const [a, b] of prerequisites) { graph[b].push(a); indegree[a]++; }
    const queue = [];
    for (let i = 0; i < numCourses; i++) if (indegree[i] === 0) queue.push(i);
    let count = 0;
    while (queue.length) {
        const course = queue.shift();
        count++;
        for (const next of graph[course]) { if (--indegree[next] === 0) queue.push(next); }
    }
    return count === numCourses;
};

File: scripts/test_solutions/207_course_schedule.py
from collections import deque

def canFinish(numCourses, prerequisites):
    graph = [[] for _ in range(numCourses)]
    indegree = [0] * numCourses
    for a, b in prerequisites:
        graph[b].append(a)
        indegree[a] += 1
    queue = deque([i for i in range(numCourses) if indegree[i] == 0])
    count = 0
    while queue:
        course = queue.popleft()
        count += 1
        for next_course in graph[course]:
            indegree[next_course] -= 1
            if indegree[next_course] == 0:
                queue.append(next_course)
    return count == numCourses
